File: hushh-adk-agents/._kushal_profile_data/Interview-Prep/Problems_By_Pattern/Hash_Map/Two_Sum/two_sum.js
// Two Sum - JavaScript implementation
// Time Complexity: O(n)
// Space Complexity: O(n)

function twoSum(nums, target) {
    const indexByValue = new Map();

    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];

        if (indexByValue.has(complement)) {
            return [indexByValue.get(complement), i];
        }

        if (!indexByValue.has(nums[i])) {
            indexByValue.set(nums[i], i);
        }
    }

    return [];
}

module.exports = { twoSum };

File: hushh-adk-agents/._kushal_profile_data/Interview-Prep/Problems_By_Pattern/Hash_Map/Two_Sum/two_sum.cs
// Two Sum - C# implementation
// Time Complexity: O(n)
// Space Complexity: O(n)

public class Solution
{
    public int[] TwoSum(int[] nums, int target)
    {
        if (nums == null || nums.Length < 2)
        {
            return Array.Empty<int>();
        }

        var indexByValue = new Dictionary<int, int>();

        for (int i = 0; i < nums.Length; i++)
        {
            int complement = target - nums[i];

            if (indexByValue.TryGetValue(complement, out int index))
            {
                return new[] { index, i };
            }

            if (!indexByValue.ContainsKey(nums[i]))
            {
                indexByValue[nums[i]] = i;
            }
        }

        return Array.Empty<int>();
    }
}
